core/src/main/scala/spinal/core/sim/SpinalSpec.scala

```diff
@@ -0,0 +1,276 @@
+      def itShould(doWhat: String): String = s"$getName should $doWhat"
+      def test(doWhat: String): String = s"$getName test: $doWhat"
+    }
+
+    /** Compile `cachedBench` if caching is enabled and not compiled yet
+      *
+      * Use this function only before tests, NEVER in a test
+      */
+    protected def prepareCachedBench(force: Boolean = false): Unit = {
+      if ((caching && !triedCompile) || force) {
+        triedCompile = true
+        txt.itCompiles in {
+          cachedBench = Some(buildBench)
+        }
+      }
+    }
+
+    /** Function used to run a test/simulation
+      *
+      * @param name The name of the test/simulation
+      * @param f doSim function, eg: `_.doSimUntilVoid(name)` or `dut => dut.doSim(name, seed)`
+      * @param body the code to run on the (possibly abstracted) `dut` during the simulation
+      */
+    protected def sim(name: String, f: SimCompiled[Dut] => (Dut => Unit) => Unit)(
+        body: Body
+    ): Unit = {
+      prepareCachedBench()
+      name in {
+        val bench =
+          if (caching) cachedBench getOrElse cancel
+          else buildBench
+        f(bench) { dut =>
+          val softDut = preSimTransform(dut)
+          body(softDut)
+        }
+      }
+    }
+
+    /** Helper to call sim with optional seed */
+    final protected def sim(name: String): Body => Unit = seed match {
+      case Some(s) => sim(name, _.doSim(name, s))
+      case None    => sim(name, _.doSim(name))
+    }
+
+    /** Helper to call sim with doSimUntilVoid and optional seed */
+    final protected def simUntilVoid(name: String): Body => Unit = seed match {
+      case Some(s) => sim(name, _.doSimUntilVoid(name, s))
+      case None    => sim(name, _.doSimUntilVoid(name))
+    }
+
+    /** Context for `it should "do something"` */
+    class Should(doWhat: String) {
+
+      /** Run a simulation to check that it does the thing it should
+        *
+        * {{{
+        * it should "do something" inSim { dut =>
+        *   // code to check that it does the thing it should
+        * }
+        * }}}
+        */
+      def inSim: Body => Unit = sim(txt.itShould(doWhat))
+
+      /** Run a simulation to check that it does the thing it should
+        *
+        * {{{
+        * it should "do something" inSimUntilVoid { dut =>
+        *   // code to check that it does the thing it should
+        * }
+        * }}}
+        */
+      def isSimUntilVoid: Body => Unit = simUntilVoid(txt.itShould(doWhat))
+    }
+
+    /** To build a [[Should]] context */
+    def should(doWhat: String): Should = new Should(doWhat)
+
+    /** `it should compile` runs a test to check that `it` compiles */
+    def should(c: Compile): Unit = prepareCachedBench(force = true)
+
+    /** `it should compile` runs a test to check that `it` does not compile */
+    def shouldNot(c: Compile): Unit = {
+      txt.notCompiles in {
+        if (caching && triedCompile) {
+          assert(cachedBench.isEmpty)
```

Comment:
maybe provide a more useful error message here, otherwise it would just state that an assertion in code that the user has not written has failed, same just below?